src/lib/network_pool/snark_pool.ml

```diff
@@ -219,6 +223,57 @@ struct
                 update_and_rebroadcast ()
               else `Don't_rebroadcast
         else `Don't_rebroadcast
+
+      let verify_proof ~(proof : Ledger_proof.t) ~statement ~message =
+        let statement_eq a b =
+          Int.(Transaction_snark.Statement.compare a b = 0)
+        in
+        if not (statement_eq (Ledger_proof.statement proof) statement) then
+          Deferred.return
+            (Or_error.error_string "Statement and proof do not match")
+        else
+          let%bind (verifier : Verifier.t) = Verifier.create () in
+          Verifier.verify_transaction_snark verifier proof ~message
+
+      let verify_and_act t ~work ~sender =
+        let statements, priced_proof = work in
+        let open Deferred.Or_error.Let_syntax in
+        let {Priced_proof.proof= proofs; fee= {prover; fee}} = priced_proof in
+        let trust_record =
+          Trust_system.record_envelope_sender t.trust_system t.logger sender
+        in
+        let log_and_punish statement e =
+          let metadata =
+            [ ("work_id", `Int (Transaction_snark.Statement.hash statement))
+            ; ("prover", Signature_lib.Public_key.Compressed.to_yojson prover)
+            ; ("fee", Currency.Fee.to_yojson fee)
+            ; ("error", `String (Error.to_string_hum e)) ]
+          in
+          Logger.error t.logger ~module_:__MODULE__ ~location:__LOC__ ~metadata
+            "Error verifying transaction snark: $error" ;
+          trust_record
+            ( Trust_system.Actions.Sent_invalid_proof
+            , Some ("Error verifying transaction snark: $error", metadata) )
+        in
+        let message = Coda_base.Sok_message.create ~fee ~prover in
+        let verify ~proof ~statement =
+          let open Deferred.Let_syntax in
+          match%bind verify_proof ~proof ~statement ~message with
+          | Ok true ->
+              Deferred.Or_error.return ()
+          | Ok false ->
+              (*Invalid proof*)
+              let e = Error.of_string "Invalid proof" in
+              let%map () = log_and_punish statement e in
+              Error e
+          | Error e ->
+              (*Other errors but punish for sending erroneous work*)
+              let%map () = log_and_punish statement e in
+              Error e
+        in
+        let%bind pairs = One_or_two.zip statements proofs |> Deferred.return in
```

Comment:
Could you use a vanilla `let` here, without the `Deferred.return`?

Reply:
it is binding a `Deferred.Or_error`